pandas/core/indexing.py

```diff
@@ -859,15 +860,22 @@ def _convert_for_reindex(self, key, axis=0):
             return labels[key]
         else:
             if isinstance(key, Index):
-                # want Index objects to pass through untouched
-                keyarr = key
+                if isinstance(labels, ABCUInt64Index) and key.is_integer():
+                    keyarr = key.astype(np.uint64)
+                else:
```

Comment:
Specially-cased, you say?  This code is WAY too `int64`-oriented that attempts to generalize the patch caused test failures all over the place.  The reason why this is needed is because when you attempt to index a `UInt64Index`, the current code implicitly does a lot of casting to `int64`, which works for about everything else except for `uint64`.